src/control/server/config/server.go

```diff
@@ -502,37 +508,112 @@ func (cfg *Server) SetNrHugepages(log logging.Logger, mi *common.MemInfo) error
 		log.Debug(msg)
 	}
 
-	if cfgTargetCount <= 0 {
-		return nil // no nvme, no hugepages required
-	}
+	return
+}
 
-	if cfg.DisableHugepages {
-		return FaultConfigHugepagesDisabledWithBdevs
+func (cfg *Server) getMinMaxNrHugepages(log logging.Logger, hpSizeKiB int) (int, int, error) {
+	cfgTargetCount, sysXSCount := cfg.getTgtCounts(log)
+
+	if cfgTargetCount == 0 {
+		return 0, 0, nil
 	}
 
 	// Calculate minimum number of hugepages for all configured engines.
-	minHugepages, err := storage.CalcMinHugepages(mi.HugepageSizeKiB, cfgTargetCount+sysXSCount)
+	minHugepages, err := storage.CalcMinHugepages(hpSizeKiB, cfgTargetCount+sysXSCount)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	maxTgtCount := largeTargetCount
+	if sysXSCount > 0 {
+		maxTgtCount += largeSysXSCount
+	}
+	maxHugepages, err := storage.CalcMinHugepages(hpSizeKiB, maxTgtCount)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	var msgSysXS string
+	if sysXSCount > 0 {
+		msgSysXS = fmt.Sprintf(" and %d/%d sys-xstreams", sysXSCount, largeSysXSCount)
+	}
+	log.Tracef("calculated min/max %d/%d nr_hugepages based on %d/%d targets%s",
+		minHugepages, maxHugepages, cfgTargetCount, largeTargetCount, msgSysXS)
+
+	if minHugepages > maxHugepages {
+		log.Debugf("config hugepage requirements exceed normal maximum")
+		maxHugepages = minHugepages
+	}
+
+	return minHugepages, maxHugepages, nil
+}
+
+// SetNrHugepages calculates minimum based on total target count if using nvme. If cfg.NrHugepages
+// is set to zero, no hugepage allocation requests will be made to the kernel during service
+// start-up in prepBdevStorage(). Only set non-zero here if a change is required.
+func (cfg *Server) SetNrHugepages(log logging.Logger, mi *common.MemInfo) error {
+	minHugepages, maxHugepages, err := cfg.getMinMaxNrHugepages(log, mi.HugepageSizeKiB)
 	if err != nil {
 		return err
 	}
 
-	// If the config doesn't specify hugepages, use the minimum. Otherwise, validate
-	// that the configured amount is sufficient.
+	// Allow emulated NVMe configurations either with or without hugepages enabled.
+
+	if cfg.DisableHugepages {
+		if cfg.NrHugepages != 0 {
+			return FaultConfigHugepagesDisabledWithNrSet
+		}
+		if cfg.GetBdevConfigs().HaveRealNVMe() {
+			return FaultConfigHugepagesDisabledWithBdevs
+		}
+		if minHugepages != 0 {
+			log.Noticef("hugepages disabled but targets will be assigned to bdevs, " +
```

Comment:
I don't disagree with doing something here, but logging "caution is advised" is not particularly helpful, IMO. Is it an error or not? What is the admin supposed to do if/when they happen to notice this message in the server log?

Reply:
this is to indicate that the server is operating in an unusual mode, the administrator should be aware of that

Reply:
If emulated NVMe can go either way, and we can't detect specifics, we could log a more specific, helpful message, for example: "Hugepages are explicitly disabled in this configuration file. Hugepages may be required depending on the underlying storage technology for the emulated SSDs. If you experience problems with this configuration, try...", etc.